sql/src/main/java/io/crate/executor/transport/StreamBucket.java

```diff
@@ -0,0 +1,192 @@
+        private final BytesStreamOutput out;
+
+        public static final Function<Builder, Bucket> BUILD_FUNCTION =
+                new Function<Builder, Bucket>() {
+                    @Nullable
+                    @Override
+                    public Bucket apply(StreamBucket.Builder input) {
+                        try {
+                            return input.build();
+                        } catch (IOException e) {
+                            Throwables.propagate(e);
+                        }
+                        return null;
+                    }
+                };
+
+        public Builder(Streamer<?>[] streamers) {
+            this.streamers = streamers;
+            this.out = new BytesStreamOutput();
+        }
+
+        public void add(Row row) throws IOException {
+            size++;
+            for (int i = 0; i < streamers.length; i++) {
+                streamers[i].writeValueTo(out, row.get(i));
+            }
+        }
+
+        public StreamBucket build() throws IOException {
+            StreamBucket sb = new StreamBucket(streamers);
+            sb.size = size;
+            sb.bytes = out.bytes();
+            return sb;
+        }
+    }
+
+    public StreamBucket(@Nullable Streamer<?>[] streamers) {
+        this.streamers = streamers;
+    }
+
+    public StreamBucket(Streamer<?>[] streamers, BytesReference bytes) {
+        this(streamers);
+        this.bytes = bytes;
+    }
+
+    @Override
+    public int size() {
+        return size;
+    }
+
+    public void streamers(Streamer<?>[] streamers) {
+        this.streamers = streamers;
+    }
+
+    public static void writeBucket(StreamOutput out, Streamer<?>[] streamers, Bucket bucket) throws IOException {
+        StreamWriter writer = new StreamWriter(out, streamers, bucket.size());
+        writer.addAll(bucket);
+    }
+
+    private class RowIterator implements Iterator<Row> {
+
+        private final StreamInput input = bytes.streamInput();
+        private int pos = 0;
+        private final Object[] current = new Object[streamers.length];
+        private final Row row = new RowN(current);
+
+        @Override
+        public boolean hasNext() {
+            return pos < size;
+        }
+
+        @Override
+        public Row next() {
+            for (int c = 0; c < streamers.length; c++) {
+                try {
+                    current[c] = streamers[c].readValueFrom(input);
+                } catch (IOException e) {
+                    Throwables.propagate(e);
+                }
+            }
+            pos++;
+            return row;
+        }
+
+        @Override
+        public void remove() {
+
+        }
+    }
+
+    @Override
+    public Iterator<Row> iterator() {
+        assert streamers != null;
+        return new RowIterator();
+    }
+
+    @Override
+    public void readFrom(StreamInput in) throws IOException {
+        size = in.readVInt();
+        BytesStreamOutput memoryStream = new BytesStreamOutput();
```

Comment:
can ``in.available()`` be used here to optimize the amount of memory allocated in BytesStreamOutput? Otherwise it will use 16KB